### PR TITLE
ic3ia -- Reduce predicates 

### PR DESCRIPTION
**Pull request by ahmed-irfan**
It uses the unsat-core-reducer from the smt-switch

### CODE REVIEW

**Review by makaimann — COMMENTED** [review]
Thanks for the feature! This looks pretty good to me, I left one minor comment and then a couple questions about a possible issue.

File: modifiers/implicit_predicate_abstractor.cpp [hunk 1]
```diff
@@ -26,9 +26,13 @@ using namespace std;
 namespace pono {
 
 ImplicitPredicateAbstractor::ImplicitPredicateAbstractor(
-    const TransitionSystem & conc_ts, TransitionSystem & abs_ts)
+    const TransitionSystem & conc_ts, TransitionSystem & abs_ts,
+    Unroller & un,
+    SmtSolver reducer_slv)
```
[the remark this answers is not in the document]
Comment (ahmed-irfan) [reply modifiers/implicit_predicate_abstractor.cpp]: good point. 
fixed in f0aa215

File: modifiers/implicit_predicate_abstractor.cpp [hunk 2]
```diff
@@ -74,13 +79,51 @@ Term ImplicitPredicateAbstractor::add_predicate(const Term & pred)
...
+  }
+
+  TermVec assumps, red_assumps;
+  UnorderedTermMap assumps_to_pred;
+  for (const auto &p : new_preds) {
+    Term pred_ref = predicate_refinement(p);
+
+    Term a;
+    for (size_t i = 0; i < cex.size(); ++i) {
+      Term p_i = unroller_.at_time(pred_ref, 0);
```
[the remark this answers is not in the document]
Comment (ahmed-irfan) [reply modifiers/implicit_predicate_abstractor.cpp]: great call :)

fixed in 7a5bd97 and d6d29df

**Comment by makaimann** [comment]
Thanks for the changes! I'll merge once Travis passes.

**Pull request merged**